packages/block-library/src/group/edit.js

```diff
@@ -67,6 +70,16 @@ function GroupEdit( { attributes, setAttributes, clientId } ) {
 		}
 	);
 
+	const { __unstableMarkNextChangeAsNotPersistent } =
+		useDispatch( blockEditorStore );
+	const { type: layoutType = null } = layout;
+	useEffect( () => {
```

Comment:
In my testing this seems to just set the layout type attribute to what it is already set to. The toggling between `constrained` and `default` seems to work as expected without this `useEffect`, but maybe there is an edge case it was added for that I am not covering.

Reply:
Yeah, this is a weird one. Though the block is inserted with the attribute, it's not saved unless we explicitly set it inside the effect. So if you add a block, save and refresh the layout type won't be there anymore. I'm not sure why this happens though 😅

Reply:
hmm, yeh, this is weird 🤔

Reply:
Perhaps this is due to Gutenberg's behavior (as I recall) of not serializing a block attribute when it's value is equal to the default? Heading blocks, for example, don't save their level when their level is set to `2`, since that's the default.

Reply:
Good point, that would explain why we need to explicitly set the attribute - we're probably checking for any `setAttribute` calls, instead of actually looking at the attributes in the markup and comparing them to defaults.

Reply:
@tellthemachines This seems to have caused a bug - https://github.com/WordPress/gutenberg/issues/43760.

I'm not sure exactly what's happening, but this effect seems to send `useBlockSync` into a frenzy in particular situations.

> we're probably checking for any setAttribute calls, instead of actually looking at the attributes in the markup and comparing them to defaults.

I think it's more that `layout` is an object type attribute, so will be referentially different to the default when set via `setAttributes`. The code that checks for default values probably only does a `===` comparison. For simple types setting to the default will result in the attribute not being output.

> So if you add a block, save and refresh the layout type won't be there anymore.

I think this is because layout code is using the 'parsed' block object, and that doesn't have defaults applied.

Defaults are applied when accessing the attributes via the `WP_Block` object. There's the magic getter here:
https://github.com/WordPress/wordpress-develop/blob/b80ba269f6cf0b0600be49b89b9a926421af88ee/src/wp-includes/class-wp-block.php#L164-L191

I think the layout code should switch to using the `WP_Block` if possible. If I'm looking at the right code, layout hooks into `render_block` and the `WP_Block` instance is available as the third param of the callback:
https://github.com/WordPress/wordpress-develop/blob/b80ba269f6cf0b0600be49b89b9a926421af88ee/src/wp-includes/class-wp-block.php#L284